ui/src/common/components/TreeAggregation.jsx

```diff
@@ -0,0 +1,164 @@
+import AggregationBox from './AggregationBox';
+import { forceArray } from '../utils';
+import FormattedNumber from './FormattedNumber';
+
+export const BUCKET_NAME_SPLITTER = '_';
+
+class TreeAggregation extends Component {
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const { selections } = nextProps;
+    const { prevSelections } = prevState;
+
+    if (selections === prevSelections) {
+      return prevState;
+    }
+
+    let selectionMap;
+    if (selections) {
+      const selectionsAsArray = forceArray(selections);
+      selectionMap = selectionsAsArray.reduce(
+        (map, key) => map.set(key, true),
+        Immutable.Map()
+      );
+    } else {
+      selectionMap = Immutable.Map();
+    }
+
+    return {
+      ...prevState,
+      prevSelections: selections,
+      selectionMap,
+    };
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      selectionMap: Immutable.Map(),
+    };
+  }
+
+  onSelectionChange(selectedKeys) {
+    let { selectionMap } = this.state;
+    const { onChange } = this.props;
+    let newSelectionMap = Immutable.Map();
+    selectedKeys.forEach(key => {
+      newSelectionMap = newSelectionMap.set(key, true);
+    });
+    selectionMap = newSelectionMap;
+    this.setState({ selectionMap });
+    const selections = selectionMap
+      .keySeq()
+      .filter(bucketKey => selectionMap.get(bucketKey))
+      .toArray();
+    onChange(selections);
+  }
+
+  getCheckedKeys(buckets) {
+    const { selectionMap } = this.state;
+    const checkedKeys = buckets
+      .map(bucket => bucket.get('key'))
+      .filter(bucketKey => selectionMap.get(bucketKey));
+    return checkedKeys.toJS();
+  }
+
+  buildTree(subTree, bucketKeyParts, part, docCount, splitTreeBy) {
+    let found = false;
+    const fullKey = bucketKeyParts.slice(0, part + 1).join(splitTreeBy);
+
+    if (part >= bucketKeyParts.length) {
+      return subTree;
+    }
+    if (subTree) {
+      for (let i = 0; i < subTree.length; i += 1) {
+        if (subTree[i].key === fullKey) {
+          found = subTree[i];
+        }
+      }
+    }
+    if (!found) {
+      const title = bucketKeyParts[part];
+      subTree.push({
+        title: (
+          <Row className="mb2" type="flex" justify="space-between" key={title}>
+            <Col>{title}</Col>
+            <Col>
+              <UnclickableTag>
+                <FormattedNumber>{docCount}</FormattedNumber>
+              </UnclickableTag>
+            </Col>
+          </Row>
+        ),
+        key: fullKey,
+        children: [],
+      });
+      // eslint-disable-next-line prefer-destructuring
+      found = subTree[0];
+    }
+    this.buildTree(
+      found.children,
+      bucketKeyParts,
+      part + 1,
+      docCount,
+      splitTreeBy
+    );
+    return subTree;
+  }
+
+  render() {
+    const { name, buckets, splitTreeBy } = this.props;
+    let tree = [];
+    for (let i = 0; i < buckets.size; i += 1) {
```

Comment:
`render` called many times so in general doing a loop like this, could lead performance problems. and that should be memoized

Reply:
also this logic should be part of `buildTree` function

Reply:
In which situations can that happen? The page is re-rendered when the filter changes and in that case we need to recreate the tree so it seems okay like this.

Reply:
you need to rebuild the tree only when `buckets` change, but this component's render method can be called for many different reasons, such other props change, internal state change, parent rerender etc.